### PR TITLE
ENG-4665 Legacy API refactoring (removed CXF)

### PR DESCRIPTION
**Pull request by zonia3000**

### CODE REVIEW

File: cms-plugin/src/main/java/com/agiletec/plugins/jacms/aps/system/services/content/model/attribute/CmsHypertextAttribute.java [hunk 1]
```diff
@@ -177,8 +177,8 @@ public List<AttributeFieldError> validate(AttributeTracer tracer, ILangManager l
                         if (null != attributeError) {
                             AttributeFieldError error = new AttributeFieldError(this, attributeError.getErrorCode(), textTracer);
                             if (attributeError.getMessage() == null) {
-                                attributeError.setMessage("Invalid link - page " + symbLink.getPageDest()
-                                        + " - content " + symbLink.getContentDest() + " - Error code " + attributeError.getErrorCode());
+                                attributeError.setMessage("Invalid link - page " + symbLink.getPageDestination()
```
Comment (entgigi) [inline cms-plugin/src/main/java/com/agiletec/plugins/jacms/aps/system/services/content/model/attribute/CmsHypertextAttribute.java]: could we use String format?

File: cms-plugin/src/main/java/com/agiletec/plugins/jacms/aps/system/services/content/model/attribute/LinkAttribute.java [hunk 2]
```diff
@@ -239,8 +239,8 @@ public List<AttributeFieldError> validate(AttributeTracer tracer, ILangManager l
             if (null != attributeError) {
                 AttributeFieldError error = new AttributeFieldError(this, attributeError.getErrorCode(), tracer);
                 if (attributeError.getMessage() == null) {
-                    attributeError.setMessage("Invalid link - page " + symbolicLink.getPageDest()
-                            + " - content " + symbolicLink.getContentDest() + " - Error code " + attributeError.getErrorCode());
+                    attributeError.setMessage("Invalid link - page " + symbolicLink.getPageDestination()
```
Comment (entgigi) [inline cms-plugin/src/main/java/com/agiletec/plugins/jacms/aps/system/services/content/model/attribute/LinkAttribute.java]: could we use String format?

File: cms-plugin/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/api/ApiContentInterface.java [hunk 3]
```diff
@@ -247,32 +246,32 @@ protected Integer checkModel(String modelId, Content content) throws ApiExceptio
                 if (null == content.getDefaultModel()) {
                     throw new ApiException(IApiErrorCodes.API_PARAMETER_VALIDATION_ERROR,
                             "Invalid 'default' system model for content type '" + content.getTypeCode() + "' - Contact the administrators",
-                            Response.Status.ACCEPTED);
+                            HttpStatus.CONFLICT);
```
Comment (entgigi) [inline cms-plugin/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/api/ApiContentInterface.java]: maybe should we use error 400?

File: cms-plugin/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/api/ApiContentTypeInterface.java [hunk 4]
```diff
@@ -155,13 +153,13 @@ private boolean checkContentModel(Integer modelId, Content contentType, StringAp
         ContentModel contentModel = this.getContentModelManager().getContentModel(modelId);
         if (null == contentModel) {
             ApiError error = new ApiError(IApiErrorCodes.API_VALIDATION_ERROR,
-					"Content model with id '" + modelId + "' does not exist", Response.Status.ACCEPTED);
+					"Content model with id '" + modelId + "' does not exist", HttpStatus.CONFLICT);
```
Comment (entgigi) [inline cms-plugin/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/api/ApiContentTypeInterface.java]: maybe should we use error 400?

File: cms-plugin/src/test/java/org/entando/entando/plugins/jacms/aps/system/services/api/TestApiContentInterface.java [hunk 5]
```diff
@@ -42,6 +42,7 @@
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.springframework.http.MediaType;
```
Comment (entgigi) [inline cms-plugin/src/test/java/org/entando/entando/plugins/jacms/aps/system/services/api/TestApiContentInterface.java:45]: should we change the name of the class ? ... ApiContentInterfaceIntegrationTest

File: cms-plugin/src/test/java/org/entando/entando/plugins/jacms/aps/system/services/api/TestApiContentTypeInterface.java [hunk 6]
```diff
@@ -16,25 +16,23 @@
...
 import org.entando.entando.plugins.jacms.aps.system.services.api.model.JAXBContentType;
-
-import javax.ws.rs.core.MediaType;
-import java.io.ByteArrayInputStream;
-import java.io.InputStream;
-import java.util.List;
-import java.util.Properties;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.springframework.http.MediaType;
```
Comment (entgigi) [inline cms-plugin/src/test/java/org/entando/entando/plugins/jacms/aps/system/services/api/TestApiContentTypeInterface.java:35]: should we change the name of the class ? ... ApiContentTypeInterfaceIntegrationTest

File: engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.List;
+import org.entando.entando.aps.system.services.api.model.ListResponse;
+import org.entando.entando.aps.system.services.api.model.ListResponseSerializer;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Primary;
+
+@Configuration
+public class ObjectMapperConfiguration {
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java:20]: i would like to use a name connected with his configuration or his use, objectMapper seems to me too generic. wWat do you think ?

File: engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+    private List<JsonTypesProvider> jsonTypesProviders;
+
+    @Autowired
+    public void setJsonTypesProviders(List<JsonTypesProvider> jsonTypesProviders) {
+        this.jsonTypesProviders = jsonTypesProviders;
+    }
+
+    @Bean("DefaultObjectMapper")
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java:29]: maybe DefaultJsonLegacyApiMapper is it a better name ?

File: engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    @Primary
+    public ObjectMapper defaultObjectMapper() {
+        ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        configureDefaultSettings(objectMapper);
+        return objectMapper;
+    }
+
+    @Bean
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java]: maybe a specific name should be use so other people are aware to not use it ... what do you think?

File: engine/src/main/java/org/entando/entando/aps/system/services/api/ApiObjectMapper.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import javax.annotation.PostConstruct;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ApiObjectMapper {
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/api/ApiObjectMapper.java]: I would like to put the word legacy in the name to warning other people to not use it ...

File: engine/src/main/java/org/entando/entando/aps/system/services/api/Unmarshaller.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.xml.XmlMapper;
+import java.io.IOException;
+import java.io.InputStream;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Component;
+
+@Component(SystemConstants.UNMARSHALLER)
+public class Unmarshaller {
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/api/Unmarshaller.java]: the name is too much generic in particular for a bean deeply coupled with legacy API  implementation, can we change it?

File: engine/src/main/java/org/entando/entando/aps/system/services/api/model/ApiError.java [hunk 12]
```diff
@@ -15,55 +15,56 @@
...
-	private String _message;
-	
-	private Response.Status _status;
-	
+
+    private String code;
+    private String message;
+    private HttpStatus status;
+
+    public ApiError() {
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/api/model/ApiError.java:21]: i fear that someone in the new API could use this so i would like change name in something like LegacyApiError if possible but it's just a suggestion ...

File: engine/src/main/java/org/entando/entando/aps/system/services/api/server/ApiRestServer.java [hunk 13]
```diff
@@ -20,153 +20,234 @@
...
-
-    @GET
-    @Produces({"application/xml", "text/plain", "application/json", "application/javascript"})
-    @Path("/{langCode}/{resourceName}")
-    public Object doGet(@PathParam("langCode") String langCode, @PathParam("resourceName") String resourceName,
-                        @Context HttpServletRequest request, @Context UriInfo ui) {
-        return this.buildGetDeleteResponse(langCode, ApiMethod.HttpMethod.GET, null, resourceName, request, ui);
+    private Unmarshaller unmarshaller;
+
+    @Autowired
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/api/server/ApiRestServer.java:62]: can we use Autowired in the class constructor?

File: engine/src/main/java/org/entando/entando/aps/system/services/api/server/ApiRestStatusServer.java [hunk 14]
```diff
@@ -65,28 +97,29 @@ public Object getApiStatus(@PathParam("httpMethod") String httpMethodString,
...
 
-    private StringApiResponse buildErrorResponse(ApiMethod.HttpMethod httpMethod,
-			String namespace, String resourceName, Throwable t) {
-		StringBuilder buffer = new StringBuilder();
-		buffer.append("Method '").append(httpMethod).
-				append("' Namespace '").append(namespace).append("' Resource '").append(resourceName).append("'");
-		_logger.error("Error building api response  - {}", buffer.toString(), t);
+    private ResponseEntity<StringApiResponse> buildErrorResponse(ApiMethod.HttpMethod httpMethod,
+            String namespace, String resourceName, Throwable t) {
+        StringBuilder buffer = new StringBuilder();
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/api/server/ApiRestStatusServer.java]: why not use a simple and more readable `String.format`?

File: engine/src/main/java/org/entando/entando/aps/system/services/guifragment/api/ApiGuiFragmentInterface.java [hunk 15]
```diff
@@ -210,15 +205,19 @@ private String getApiResourceUrl(String fragmentCode, String applicationBaseUrl,
 			return null;
 		}
 		StringBuilder stringBuilder = new StringBuilder(applicationBaseUrl);
-		stringBuilder.append("api/rs/").append(langCode).append("/core/guiFragment");//?code=").append(fragmentCode);
-		if (null == mediaType || mediaType.equals(MediaType.APPLICATION_XML_TYPE)) {
+		stringBuilder.append("api/legacy/").append(langCode).append("/core/guiFragment");
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/guifragment/api/ApiGuiFragmentInterface.java]: i'm really afraid to put this a String in the code ... can we use a common constant shared with the controller configuration

File: engine/src/main/java/org/entando/entando/aps/system/services/pagemodel/api/ApiPageModelInterface.java [hunk 16]
```diff
@@ -163,8 +158,8 @@ public String getApiResourceUrl(Object object, String applicationBaseUrl, String
 		}
 		PageModel pageModel = (PageModel) object;
 		StringBuilder stringBuilder = new StringBuilder(applicationBaseUrl);
-		stringBuilder.append("api/rs/").append(langCode).append("/core/pageModel");//?code=").append(pageModel.getCode());
-		if (null == mediaType || mediaType.equals(MediaType.APPLICATION_XML_TYPE)) {
+		stringBuilder.append("api/legacy/").append(langCode).append("/core/pageModel");
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/pagemodel/api/ApiPageModelInterface.java]: can we use a constant for `legacy` or `/core/pageModel` (shared with controller configuration i mean ...) ?

File: engine/src/main/java/org/entando/entando/aps/system/services/storage/api/LocalStorageManagerInterface.java [hunk 17]
```diff
@@ -213,15 +213,15 @@ public String getApiResourceUrl(Object object, String applicationBaseUrl, String
 		}
 		BasicFileAttributeView basicFileAttributeView = (BasicFileAttributeView) object;
 		StringBuilder stringBuilder = new StringBuilder(applicationBaseUrl);
-		stringBuilder.append("api/rs/").append(langCode).append("/core/");
+		stringBuilder.append("api/legacy/").append(langCode).append("/core/");
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/storage/api/LocalStorageManagerInterface.java]: shared constants ....

File: engine/src/main/java/org/entando/entando/aps/system/services/widgettype/api/ApiWidgetTypeInterface.java [hunk 18]
```diff
@@ -310,8 +305,8 @@ public String getApiResourceUrl(Object object, String applicationBaseUrl, String
 		}
 		WidgetType widgetType = (WidgetType) object;
 		StringBuilder stringBuilder = new StringBuilder(applicationBaseUrl);
-		stringBuilder.append("api/rs/").append(langCode).append("/core/widgetType");// ?code=").append(widgetType.getCode());
-		if (null == mediaType || mediaType.equals(MediaType.APPLICATION_XML_TYPE)) {
+		stringBuilder.append("api/legacy/").append(langCode).append("/core/widgetType");
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/widgettype/api/ApiWidgetTypeInterface.java]: shared constants ...

File: engine/src/main/java/org/entando/entando/aps/util/UrlUtils.java [hunk 19]
```diff
@@ -218,11 +219,15 @@ public EntUrlBuilder paths(String ... u){
             return this;
         }
 
-        public URI build(){
-            UriBuilder builder = UriBuilder.fromUri(url);
+        public URI build() {
+            UriBuilder builder = UriComponentsBuilder.fromHttpUrl(url);
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/util/UrlUtils.java]: has this method the same behavior? (from uri from url sound strange to me ...)

**Review by entgigi — CHANGES_REQUESTED** [review]

File: engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java (hunk 8 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java:29]: This is used by all the API endpoints, not only the legacy ones.
It replaces this bean that was defined using XML syntax:
https://github.com/entando/app-engine/blob/6b85f99ccf6ce37d360855327e63570d4819c48f/engine/src/main/resources/spring/aps/servicesConfig.xml#L164

Comment (entgigi) [reply engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java:29]: ok so we cannot touch it, I got it

File: engine/src/main/java/org/entando/entando/aps/system/services/api/ApiObjectMapper.java (hunk 10 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/api/ApiObjectMapper.java]: This bean is not used, I remove it

File: cms-plugin/src/main/java/com/agiletec/plugins/jacms/aps/system/services/content/model/attribute/CmsHypertextAttribute.java (hunk 1 above)
Comment (zonia3000) [reply cms-plugin/src/main/java/com/agiletec/plugins/jacms/aps/system/services/content/model/attribute/CmsHypertextAttribute.java]: done

File: cms-plugin/src/main/java/com/agiletec/plugins/jacms/aps/system/services/content/model/attribute/LinkAttribute.java (hunk 2 above)
Comment (zonia3000) [reply cms-plugin/src/main/java/com/agiletec/plugins/jacms/aps/system/services/content/model/attribute/LinkAttribute.java]: done

File: cms-plugin/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/api/ApiContentInterface.java (hunk 3 above)
Comment (zonia3000) [reply cms-plugin/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/api/ApiContentInterface.java]: done

File: cms-plugin/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/api/ApiContentTypeInterface.java (hunk 4 above)
Comment (zonia3000) [reply cms-plugin/src/main/java/org/entando/entando/plugins/jacms/aps/system/services/api/ApiContentTypeInterface.java]: done

File: cms-plugin/src/test/java/org/entando/entando/plugins/jacms/aps/system/services/api/TestApiContentInterface.java (hunk 5 above)
Comment (zonia3000) [reply cms-plugin/src/test/java/org/entando/entando/plugins/jacms/aps/system/services/api/TestApiContentInterface.java:45]: done

File: cms-plugin/src/test/java/org/entando/entando/plugins/jacms/aps/system/services/api/TestApiContentTypeInterface.java (hunk 6 above)
Comment (zonia3000) [reply cms-plugin/src/test/java/org/entando/entando/plugins/jacms/aps/system/services/api/TestApiContentTypeInterface.java:35]: done

File: engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java (hunk 7 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java:20]: As said above, the DefaultObjectMapper is used by all the API endpoints.

File: engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java (hunk 9 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/api/ObjectMapperConfiguration.java]: Named `LegacyApiXmlMapper`

File: engine/src/main/java/org/entando/entando/aps/system/services/api/Unmarshaller.java (hunk 11 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/api/Unmarshaller.java]: Named `LegacyApiUnmarshaller`. Also renamed the constant.

File: engine/src/main/java/org/entando/entando/aps/system/services/api/model/ApiError.java (hunk 12 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/api/model/ApiError.java:21]: Renamed to `LegacyApiError`.

File: engine/src/main/java/org/entando/entando/aps/system/services/api/server/ApiRestServer.java (hunk 13 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/api/server/ApiRestServer.java:62]: done

File: engine/src/main/java/org/entando/entando/aps/system/services/api/server/ApiRestStatusServer.java (hunk 14 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/api/server/ApiRestStatusServer.java]: done

File: engine/src/main/java/org/entando/entando/aps/system/services/api/ApiObjectMapper.java (hunk 10 above)
Comment (entgigi) [reply engine/src/main/java/org/entando/entando/aps/system/services/api/ApiObjectMapper.java]: great

File: engine/src/main/java/org/entando/entando/aps/system/services/guifragment/api/ApiGuiFragmentInterface.java (hunk 15 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/guifragment/api/ApiGuiFragmentInterface.java]: I created `SystemConstants.LEGACY_API_PREFIX`, referenced both in these classes and in the controller. I've also replaced the `StringBuilder` with a `String.format()` and moved the extraction of the extension in a common method.

File: engine/src/main/java/org/entando/entando/aps/util/UrlUtils.java (hunk 19 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/util/UrlUtils.java]: I used `fromUriString`, but we are passing an URL in any case.

File: engine/src/main/java/org/entando/entando/aps/system/services/widgettype/api/ApiWidgetTypeInterface.java (hunk 18 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/widgettype/api/ApiWidgetTypeInterface.java]: done, as above

File: engine/src/main/java/org/entando/entando/aps/system/services/storage/api/LocalStorageManagerInterface.java (hunk 17 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/storage/api/LocalStorageManagerInterface.java]: done, as above

File: engine/src/main/java/org/entando/entando/aps/system/services/pagemodel/api/ApiPageModelInterface.java (hunk 16 above)
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/pagemodel/api/ApiPageModelInterface.java]: done, as above

**Review by entgigi — APPROVED** [review]
LGTM

**Review by eugeniosant — APPROVED** [review]

**Pull request merged**